### PR TITLE
initial version of dedicated client scope

### PR DESCRIPTION
**Pull request by edewit**

### CODE REVIEW

File: src/clients/scopes/ClientScopes.tsx [hunk 1]
```diff
@@ -110,11 +118,27 @@ export const ClientScopes = ({
 
     const filter =
       searchType === "name" ? nameFilter(search) : typeFilter(searchTypeType);
-    return rows.filter(filter).slice(first, Number(first) + Number(max));
+    const firstNum = Number(first);
+    const page = rows
+      .filter(filter)
+      .slice(firstNum, Number(first) + Number(max));
```
Comment (jonkoops) [inline src/clients/scopes/ClientScopes.tsx:124]: You already have `firstNum` here which has been converted to a number.
```suggestion
      .slice(firstNum, firstNum + Number(max));
```

File: src/clients/scopes/DecicatedScope.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    const [assignedRoles, effectiveRoles] = await Promise.all([
+      adminClient.clients
+        .listRealmScopeMappings({ id: client.id! })
+        .then((roles) => roles.map((role) => ({ role }))),
+      adminClient.clients
+        .listCompositeRealmScopeMappings({ id: client.id! })
+        .then((roles) => roles.map((role) => ({ role }))),
+    ]);
+
+    const clients = await adminClient.clients.find();
```
Comment (jonkoops) [inline src/clients/scopes/DecicatedScope.tsx:47]: I think you can also get the clients in parallel here.
```suggestion
    const [clients, assignedRoles, effectiveRoles] = await Promise.all([
      adminClient.clients.find(),
      adminClient.clients
        .listRealmScopeMappings({ id: client.id! })
        .then((roles) => roles.map((role) => ({ role }))),
      adminClient.clients
        .listCompositeRealmScopeMappings({ id: client.id! })
        .then((roles) => roles.map((role) => ({ role }))),
    ]);

```

File: src/clients/scopes/DecicatedScope.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+          );
+          return mapRoles(clientAssignedRoles, clientEffectiveRoles, hide);
+        })
+      )
+    ).flat();
+
+    return [...mapRoles(assignedRoles, effectiveRoles, hide), ...clientRoles];
+  };
+
+  const assignRoles = async (rows: Row[]) => {
```
Comment (jonkoops) [inline src/clients/scopes/DecicatedScope.tsx:75]: Perhaps the calls in this function can be executed in parallel?

File: src/clients/scopes/DecicatedScope.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+export const DedicatedScope = ({
+  client: initialClient,
+}: DedicatedScopeProps) => {
+  const { t } = useTranslation("clients");
+  const adminClient = useAdminClient();
+  const { addAlert, addError } = useAlerts();
+
+  const [client, setClient] = useState<ClientRepresentation>(initialClient);
+  const [hide, setHide] = useState(false);
```
Comment (jonkoops) [inline src/clients/scopes/DecicatedScope.tsx:35]: `useToggle` was giving you issues here?

File: src/components/role-mapping/AddRoleMappingModal.tsx [hunk 5]
```diff
@@ -72,37 +78,15 @@ export const AddRoleMappingModal = ({
...
-                );
-                break;
+            roles = (await castAdminClient(
+              adminClient,
+              mapType.resource === "roles" ? "clients" : mapType.resource
+            )[mapType.functions.list[0]]({
+              id: mapType.resource === "roles" ? client.id : id,
+              clientUniqueId: client.id,
+              client: client.id,
+            })) as RoleRepresentation[];
```
Comment (jonkoops) [inline src/components/role-mapping/AddRoleMappingModal.tsx:88]: Is this type missing somewhere? Can we get rid of this type-cast?

File: src/components/role-mapping/AddRoleMappingModal.tsx [hunk 6]
```diff
@@ -138,32 +122,12 @@ export const AddRoleMappingModal = ({
...
-        availableRoles =
-          await adminClient.clientScopes.listAvailableRealmScopeMappings({
-            id,
-          });
-        break;
-      case "role":
-        availableRoles = await adminClient.roles.find();
-        break;
-    }
+    const availableRoles: RoleRepresentation[] = await castAdminClient(
```
Comment (jonkoops) [inline src/components/role-mapping/AddRoleMappingModal.tsx:125]: Here also, why does `castAdminClient` not have this type already?

File: src/components/role-mapping/AddRoleMappingModal.tsx [hunk 7]
```diff
@@ -181,35 +145,15 @@ export const AddRoleMappingModal = ({
...
-                  { id, client: client.id! }
-                );
-              break;
-            case "role":
-              clientAvailableRoles = await adminClient.clients.listRoles({
-                id: client.id!,
-              });
-              break;
-          }
+          const clientAvailableRoles: RoleRepresentation[] =
```
Comment (jonkoops) [inline src/components/role-mapping/AddRoleMappingModal.tsx:148]: And here as well.

File: src/components/role-mapping/RoleMapping.tsx [hunk 8]
```diff
@@ -72,15 +77,87 @@ export const ServiceRole = ({ role, client }: Row) => (
...
+  {
+    resource: "roles",
+    functions: {
+      delete: [],
+      list: ["listRoles", "find"],
+    },
+  },
+];
+
+export const castAdminClient = (
```
Comment (jonkoops) [inline src/components/role-mapping/RoleMapping.tsx:153]: Can we give this function a proper return type somehow?

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: src/clients/scopes/DecicatedScope.tsx (hunk 4 above)
Comment (edewit) [reply src/clients/scopes/DecicatedScope.tsx]: no this is code that has been implemented before `useToggle` was introduced, changed it now

File: src/client-scopes/ClientScopesSection.tsx [hunk 9]
```diff
@@ -140,7 +141,7 @@ export default function ClientScopesSection() {
       all
       onSelect={async (value) => {
```
[the remark this answers is not in the document]
Comment (edewit) [reply src/client-scopes/ClientScopesSection.tsx:142]: onSelect gives us a `string | SelectOptionObject` we know it's a string and one of `AllClientScopeType`, but we can't enforce that here

File: src/clients/ClientDetails.tsx [hunk 10]
```diff
@@ -449,23 +400,6 @@ export default function ClientDetails() {
                 <Credentials clientId={clientId} save={() => save()} />
               </Tab>
             )}
-            {!isRealmClient(client) && (
-              <Tab
-                id="mappers"
-                data-testid="mappersTab"
```
[the remark this answers is not in the document]
Comment (edewit) [reply src/clients/ClientDetails.tsx:455]: done

**Comment by jonkoops** [comment]
All tests are passing :heavy_check_mark: 
```log
  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  authentication_policies.spec.ts          00:15        3        3        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  authentication_test.spec.ts              01:09       15       15        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  client_authorization_test.spec.ts        00:22        9        9        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  client_scopes_test.spec.ts               00:20        4        4        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  clients_saml_test.spec.ts                00:15        5        5        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  clients_test.spec.ts                     00:51       15       15        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  events_test.spec.ts                      00:49       14       14        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  group_test.spec.ts                       00:35        9        9        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  identity_providers_oidc_test.spec.t      00:26        5        5        -        -        - │
  │    s                                                                                           │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  identity_providers_saml_test.spec.t      01:02       12       12        -        -        - │
  │    s                                                                                           │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  identity_providers_test.spec.ts          00:42        9        9        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  login_test.spec.ts                       00:06        2        2        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  masthead_test.spec.ts                    00:47        2        2        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  partial_export_test.spec.ts              00:12        3        3        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  partial_import_test.spec.ts              00:36        4        4        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_roles_test.spec.ts                 00:22        3        3        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_settings_client_policies_test      00:46       18       18        -        -        - │
  │    .spec.ts                                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_settings_client_profiles_test      00:43       21       21        -        -        - │
  │    .spec.ts                                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_settings_events_test.spec.ts       01:17       17       16        -        1        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_settings_tabs_test.spec.ts         00:24        6        6        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_settings_user_profile_tab.spe      00:13        3        3        -        -        - │
  │    c.ts                                                                                        │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_test.spec.ts                       00:10        4        4        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  realm_user_registration.spec.ts          00:08        3        3        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  sessions_test.spec.ts                    00:34       11       11        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  user_fed_kerberos_test.spec.ts           00:44        9        9        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  user_fed_ldap_hardcoded_mapper_test      01:15       12       12        -        -        - │
  │    .spec.ts                                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  user_fed_ldap_mapper_test.spec.ts        01:21       12       12        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  user_fed_ldap_test.spec.ts               00:50        8        8        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  users_test.spec.ts                       00:26       13       12        -        1        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        17:52      251      249        -        2        -  
```

**Review by jonkoops — APPROVED** [review]

**Pull request merged**